File: cpp/jump_game_ii.cc
/*
 * Filename: jump_game_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 26 Jul 2015 11:21:41 AM EDT
 */

#include <iostream>
#include <climits>

using namespace std;

class Solution {
	public:
		int jump(vector<int>& nums) {
			if (nums.size() < 2) {
				return 1;
			}

			int stt = 0;
			int end = 0;
			int eend = 0;
			int cnt = 0;

			while (end < nums.size() - 1) {
				eend = end;
				cnt++;

				for (vector<int>::size_type i = stt; i <= end; i++) {
					if (i + nums[i] >= n-1) {
						return cnt;
					}

					eend = max(eend, i + nums[i]);
				}

				stt = end + 1;
				end = eend;
			}
		}
};

File: java/JumpGameII.java
/*
 * Filename: JumpGameII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 11 May 2015 03:50:18 PM EDT
 */

public class Solution {
	public int jump(int[] nums) {
		if (nums == null || nums.length == 0) {
			return 0;
		}

		int curr = 0;
		int ret;
		int max;
		int idx = 0;
		int val;

		if (nums.length == 1) {
			ret = 0;
		} else {
			ret = 1;
		}

		while (curr + nums[curr] < nums.length - 1) {
			max = 0;
			for (int j = curr + 1; j <= curr + nums[curr]; j++) {
				val = j + nums[j];
				if (max < val) {
					max = val;
					idx = j;
				}
			}
			curr = idx;
			ret++;
		}

		return ret;
	}
}
